core/src/main/java/io/questdb/cairo/mv/MatViewRefreshJob.java

```diff
@@ -113,6 +118,35 @@ public boolean run(int workerId, @NotNull RunStatus runStatus) {
         return processNotifications();
     }
 
+    private static long approxPartitionMicros(int partitionBy) {
+        switch (partitionBy) {
+            case PartitionBy.HOUR:
+                return Timestamps.HOUR_MICROS;
+            case PartitionBy.DAY:
+                return Timestamps.DAY_MICROS;
+            case PartitionBy.WEEK:
+                return Timestamps.WEEK_MICROS;
+            case PartitionBy.MONTH:
+                return 31 * Timestamps.DAY_MICROS;
+            case PartitionBy.YEAR:
+                return 365 * Timestamps.DAY_MICROS;
+            default:
+                throw new UnsupportedOperationException("unexpected partition by: " + partitionBy);
+        }
+    }
+
+    /**
+     * Estimates density of rows per SAMPLE BY bucket. The estimate is not very precise as
+     * it doesn't use exact min/max timestamps for each partition, but it should do the job
+     * of splitting large refresh table scans into multiple smaller scans.
+     */
+    private static long estimateRowsPerBucket(@NotNull TableReader baseTableReader, long bucketMicros) {
+        final long rows = baseTableReader.size();
+        final long partitionMicros = approxPartitionMicros(baseTableReader.getPartitionedBy());
+        final int partitionCount = baseTableReader.getPartitionCount();
```

Comment:
assert partitionCount != 0

Reply:
Resolved in 069557f

Reply:
Tweaked in f428f7a9115e26b0438bcb4e5ffde8bad1440aa4

Not everything deserves to be an assertion